lib/ClangImporter/ImportDecl.cpp

```diff
@@ -3352,15 +3352,38 @@ namespace {
         }
       }
 
+      bool retunrsRetainedUnretainedDiagnosticsNeeded = false;
+      if (!isa<clang::CXXDeductionGuideDecl>(decl)) {
+        if (const auto *methodDecl = dyn_cast<clang::CXXMethodDecl>(decl)) {
+          if (isNormalCxxMemberMethod(methodDecl)) {
+            // normal c++ member method
+            retunrsRetainedUnretainedDiagnosticsNeeded = true;
+          }
+        } else {
+          // global or top-level function
+          retunrsRetainedUnretainedDiagnosticsNeeded = true;
+        }
+      }
+
       HeaderLoc loc(decl->getLocation());
       if (isForeignReferenceTypeWithoutImmortalAttrs(decl->getReturnType())) {
         if (returnsRetainedAttrIsPresent && returnsUnretainedAttrIsPresent) {
           Impl.diagnose(loc, diag::both_returns_retained_returns_unretained,
                         decl);
         } else if (!returnsRetainedAttrIsPresent &&
                    !returnsUnretainedAttrIsPresent) {
-          Impl.diagnose(loc, diag::no_returns_retained_returns_unretained,
-                        decl);
+          if (retunrsRetainedUnretainedDiagnosticsNeeded) {
+            if (Impl.isCxxInteropCompatVersionAtLeast(
+                    version::getUpcomingCxxInteropCompatVersion())) {
+              // TODO: change into error here
+              Impl.diagnose(loc,
+                            diag::no_returns_retained_returns_unretained_error,
```

Comment:
If the only difference is the diagnostic code, maybe you can simplify a bit by selecting the right diag into a variable and calling `diagnose` once.

Reply:
Can do this refactoring later.

Reply:
Maybe we should leave it as it is if this is a minor issue. The current way code looks clearer to me to have different diagnostic calls inside different if conditions.